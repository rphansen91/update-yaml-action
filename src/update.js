"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateYamlAction = exports.updateYaml = void 0;
const jsonpath_1 = __importDefault(require("jsonpath"));
const yaml_1 = require("yaml");
const octokit_1 = require("./octokit");
function updateYaml(content, propertyPath, value) {
    const data = (0, yaml_1.parse)(content);
    jsonpath_1.default.value(data, propertyPath, value);
    return (0, yaml_1.stringify)(data);
}
exports.updateYaml = updateYaml;
function updateYamlAction(action) {
    return __awaiter(this, void 0, void 0, function* () {
        const octo = new octokit_1.AxiosOctokit({
            token: action.token,
            repo: action.repo,
        });
        const content = yield octo.loadContents(action.valueFile, action.ref);
        const updated = updateYaml(content, action.propertyPath, action.value);
        const { url, ref } = yield octo.uploadToRepo(action.valueFile, updated, action.msg, action.ref);
        return { url, ref };
    });
}
exports.updateYamlAction = updateYamlAction;
