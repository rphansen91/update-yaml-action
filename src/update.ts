import jp from 'jsonpath'
import { parse, stringify } from 'yaml'
import { AxiosOctokit } from './octokit'

export interface ActionOptions {
    repo: string
    ref: string
    msg: string
    valueFile: string
    propertyPath: string
    value: string
    token: string
}

export function updateYaml(
  content: string,
  propertyPath: string,
  value: string
) {
  const data = parse(content);

  jp.value(data, propertyPath, value);

  return stringify(data);
}

export async function updateYamlAction(action: ActionOptions) {
  const octo = new AxiosOctokit({
    token: action.token,
    repo: action.repo,
  });

  const content = await octo.loadContents(action.valueFile, action.ref);

  const updated = updateYaml(content, action.propertyPath, action.value);

  const { url, ref } = await octo.uploadToRepo(
    action.valueFile,
    updated,
    action.msg,
    action.ref
  );

  return { url, ref };
}
