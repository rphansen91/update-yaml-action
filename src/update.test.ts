import { updateYaml } from './update'

const test = `deployments:
- name: api-coin-manager-apollo
  image:
    repository: "gcr.io/xyo-network-1522800011804/api-coin-manager-apollo"
    tag: "a47c269411afdaa935ec5e7948591c979a192b69"
  env:
    - name: BASE
      value: "/manager"`

describe('Update YAML', () => {
  it.each([
    [test, "$.deployments[?(@.name=='api-coin-manager-apollo')].image.tag", '123']
  ])('Should set nested array field', (a, b, c) => {
    expect(updateYaml(a, b, c)).toMatchSnapshot()
  })
})