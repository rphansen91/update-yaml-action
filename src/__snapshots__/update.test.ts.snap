// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update YAML Should set nested array field 1`] = `
"deployments:
  - name: api-coin-manager-apollo
    image:
      repository: gcr.io/xyo-network-1522800011804/api-coin-manager-apollo
      tag: "123"
    env:
      - name: BASE
        value: /manager
"
`;
